---
title: "C. elegans data analysis: KEGG Analysis (clusterProfiler)"
author: "Chinaza"
date: "`r Sys.Date()`"
format: html
editor: visual
embed-resources: true
---

## Setting working directory and loading packages

```{r}
#| label: setup

knitr::opts_knit$set(root.dir = "G:/My Drive/PhD/project/project_with_Dr.Reid/c.elegans_RNAseq_suppl/data_analysis/data_and_result/")


lst = c("clusterProfiler", "tibble", "pathview", "ggplot2")

load_libraries = function(list_of_packages){
  for (i in lst) {
    suppressPackageStartupMessages(library(i, character.only = TRUE))}
}

load_libraries(lst)

```

## Function

```{r}

set.seed(1234)

KEGG_ORA_n_GSEA_analysis = function(path){

        setwd("G:/My Drive/PhD/project/project_with_Dr.Reid/c.elegans_RNAseq_suppl/data_analysis/data_and_result/")
        
        modify_plot = theme(plot.title = element_text(size = 24, face ="bold", hjust = 0.5))
        
        df = read.table(path, header=TRUE, sep="\t")
        df = rownames_to_column(df, var = "ncbi_locus_tag")
        print("DeSeq2 result")
        print(head(df))
        cat("\n\n")
        
        ids = read.csv("DAVID-locus_tag_to_entrez_id.csv")
        names(ids)[names(ids) == "From"] = "ncbi_locus_tag"
        names(ids)[names(ids) == "To"] = "Entrez_id"
        print("Locus tag to entrez ID")
        print(head(ids))
        cat("\n\n")
        
        merged_df = merge(df, ids, by = "ncbi_locus_tag", all = TRUE)
        print("Merged dataframe")
        print(head(merged_df))
        cat("\n\n")
        
        genes_of_interest = as.character((subset(merged_df, padj < 0.05 & abs(log2FoldChange) > 1))$ncbi_locus_tag)
        
                                          
        # KEGG pathway: over-representation analysis using Fisher's test
        kk_ora = enrichKEGG(gene         = genes_of_interest,
                        organism     = 'cel',
                        universe      = merged_df$ncbi_locus_tag,
                        pvalueCutoff = 0.05)
        print("over-representation analysis result")
        result_kk_ora = data.frame(kk_ora)
        print(result_kk_ora)
        cat("\n\n")
        
        if ((dim(result_kk_ora)[1]) > 0){
          print(dotplot(kk_ora, showCategory = dim(result_kk_ora)[1]) + 
            labs(title="Enriched Pathway (ORA)", 
                 caption="GeneRatio ==> ratio of differential expressed genes that are annotated in a pathway\nCount ==> GeneCount") + 
            modify_plot)
        }
        
        # prepare geneList for GSEA
        geneList = merged_df$log2FoldChange
        names(geneList) = merged_df$ncbi_locus_tag
        geneList = sort(geneList, decreasing = TRUE)
        
        
        # KEGG pathway: GSEA using KS test
        kk_gsea = gseKEGG(geneList     = geneList,
                       organism     = 'cel',
                       pvalueCutoff = 0.05,
                       verbose      = FALSE,
                       seed = TRUE)
        print("GSEA analysis result")
        result_kk_gsea = data.frame(kk_gsea)
        print(result_kk_gsea)
        cat("\n\n")
        
        if ((dim(result_kk_gsea)[1]) > 0){
          print(dotplot(kk_gsea, showCategory = dim(result_kk_gsea)[1]) + 
            labs(title="Enriched Pathway (GSEA)", 
                 caption="GeneRatio ==> ratio of differential expressed genes that are annotated in a pathway\nCount ==> GeneCount") + 
            modify_plot)
        }
}

```

## Control vs TPMG

```{r fig.height=8, fig.width=8}

path = "differential_expression_analysis/deseq2/tpmg_vs_control_df.txt"

KEGG_ORA_n_GSEA_analysis(path)
```

## Control vs TYFD

```{r fig.height=8, fig.width=8}

path = "differential_expression_analysis/deseq2/tyfd_vs_control_df.txt"

KEGG_ORA_n_GSEA_analysis(path)
```

## Control vs TYEX

```{r fig.height=8, fig.width=8}

path = "differential_expression_analysis/deseq2/tyex_vs_control_df.txt"

KEGG_ORA_n_GSEA_analysis(path)
```

## Session info

```{r}

sessionInfo()
```
